src/object/cli_obj.c

```diff
@@ -1530,49 +1470,496 @@ obj_shard_task_sched(struct obj_auxi_args *obj_auxi)
+	D_ASSERT(shard_auxi->grp_idx < req_tgts->ort_grp_nr);
+	fw_shard_tgts = req_tgts->ort_srv_disp ?
+			(req_tgts->ort_shard_tgts +
+			 shard_auxi->grp_idx * req_tgts->ort_grp_size + 1) :
+			 NULL;
+	fw_cnt = req_tgts->ort_srv_disp ? (req_tgts->ort_grp_size - 1) : 0;
+	rc = shard_auxi->shard_io_cb(obj_shard, obj_auxi->opc, shard_auxi,
+				     fw_shard_tgts, fw_cnt, task);
+	obj_shard_close(obj_shard);
+
+	return rc;
+}
+
+static int
+shard_io_task(tse_task_t *task)
+{
+	struct shard_auxi_args		*shard_auxi;
+
+	shard_auxi = tse_task_buf_embedded(task, sizeof(*shard_auxi));
+
+	return shard_io(task, shard_auxi);
+}
+
+typedef int (*shard_io_prep_cb_t)(struct shard_auxi_args *shard_auxi,
+				  struct dc_object *obj,
+				  struct obj_auxi_args *obj_auxi,
+				  uint64_t dkey_hash, uint32_t map_ver,
+				  daos_epoch_t epoch);
+
+static int
+shard_task_reset_target(tse_task_t *shard_task, void *arg)
+{
+	struct obj_req_tgts		*req_tgts = arg;
+	struct obj_auxi_args		*obj_auxi;
+	struct shard_auxi_args		*shard_arg;
+	struct daos_obj_shard_tgt	*leader_tgt;
+
+	shard_arg = tse_task_buf_embedded(shard_task, sizeof(*shard_arg));
+	D_ASSERT(shard_arg->grp_idx < req_tgts->ort_grp_nr);
+	obj_auxi = container_of(req_tgts, struct obj_auxi_args, req_tgts);
+	leader_tgt = req_tgts->ort_shard_tgts +
+		     shard_arg->grp_idx * req_tgts->ort_grp_size;
+	shard_arg->shard = leader_tgt->st_shard;
+	shard_arg->target = leader_tgt->st_tgt_id;
+	shard_arg->map_ver = obj_auxi->map_ver_req;
+
+	return 0;
+}
+
+static int
+obj_req_fanout(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
+	       uint64_t dkey_hash, uint32_t map_ver, daos_epoch_t epoch,
+	       shard_io_prep_cb_t io_prep_cb, shard_io_cb_t io_cb,
+	       tse_task_t *obj_task)
+{
+	struct obj_req_tgts		*req_tgts = &obj_auxi->req_tgts;
+	d_list_t			*task_list = &obj_auxi->shard_task_head;
+	tse_task_t			*shard_task;
+	struct shard_auxi_args		*shard_auxi;
+	struct daos_obj_shard_tgt	*tgt;
+	uint32_t			 i, tgts_nr;
+	int				 rc = 0;
+
+	/* for retried obj IO, reuse the previous shard tasks and resched it */
+	if (obj_auxi->io_retry) {
+		/* We mark the RPC as RESEND although @io_retry does not
+		 * guarantee that the RPC has ever been sent. It may cause
+		 * some overhead on server side, but no correctness issues.
+		 *
+		 * On the other hand, the client may resend the RPC to new
+		 * shard if leader switched. That is why the resend logic
+		 * is handled at object layer rather than shard layer.
+		 */
+		obj_auxi->flags = ORF_RESEND;
+
+		/* if with shard task list, reuse it and re-schedule */
+		if (!d_list_empty(task_list)) {
+			/* For srv dispatch, the task_list non-empty is only for
+			 * obj punch that with multiple RDG that each with a
+			 * leader. Here reset the header for the shard task.
+			 */
+			if (req_tgts->ort_srv_disp) {
+				D_ASSERT(obj_auxi->opc == DAOS_OBJ_RPC_PUNCH);
+				tse_task_list_traverse(task_list,
+					shard_task_reset_target, req_tgts);
```

Comment:
Ah, so you reset the target here. So do you still need check target in shard_task_sched() again? I mean here                 /* For retried IO, check if the shard's target changed after
                 * pool map query. If match then need not do anything, if
                 * mismatch then need to re-schedule the shard IO on the new
                 * pool map.
                 * Also retry the shard IO if it got retryable error last time.
                 */

Reply:
Now the functionality should be fine, as that logic in shard_task_sched() can work for cli-replica, here just reset the leader replica's shard/tgt.
I agree we probably can make it cleaner, I plan to change that in following patch.